### ISSUE TITLE
Base camp expansion buildings generating without roofs

### ISSUE THREAD

**Issue by PatrikLundell** [CleverRaven/Cataclysm-DDA#70603]
### Describe the bug

For some reason base camp expansion and base camp buildings sometimes generate without roofs.

### Attach save file

N/A. The compressed saves are too big to be uploaded.

### Steps to reproduce

1. Generate a base camp.
2. Generate a livestock version 2 expansion in the NW or E expansion directions (direction probably doesn't matter).
3. Collect materials for a log walled chicken coop and order companions to build it.
4. (debug) Wait until done, walk to it, grab a (debug spawned) step ladder and place it by the wall.
5. Try to climb the ladder, and see that you can't.

### Expected behavior

Base camp construction of walls and rooms with roofs should have the roofs generated when constructed.

### Screenshots

You can't see that the roofs are missing from the ground level. You only realize it if you try to climb the building (and fail) or get up to a nearby roof and realize you see the same thing as you do from the ground level (with some height indicating fog).

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.3803 (22H2)
- Game Version: 0.G-6046-g8f6b2d45fa-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

I've tried to investigate, and this is what I've found:
- construction.cpp operation complete_construction (used when constructing using a construction zone) has code checking for whether a roof should be added, and, if so, adds it to the Z level above.
- map.cpp operation map::ter_set (used by base camp construction) does not have any corresponding logic.
- When trying to copy the logic from construction.cpp to map.cpp the added code gets thwarted because the operation unsafe_get_submap_at() (either called directly to get a submap of the level above, or indirectly by trying to use ter_set recursively (direct translation from construction.cpp) screws you over by silently returning the map of the current Z level rather than the one above because some class variable called zlevels is false.
- The class variable zlevels does not seem to be related to the experimental Z level view (the modified code returns data for Z level 0 regardless).

map.cpp operation ter_set might not be the appropriate place to add roofs. It's possible a caller higher up in the call chain should have an additional pass to add roofs after the floors/walls/... have all been added. After all, there are lots of passes where stuff in added one layer after another.

Also note that roofs show up sometimes, when constructing base camp buildings (it was some time since I last saw it due to a lack of building generation, but I have no reason to believe the situation has changed), but I have no idea why. As indicated by the reproduction steps, I've tried to build my sample building both without any rotation/mirroring and rotated/mirrored, and both versions fail.

**Comment by PatrikLundell** [comment]
No base camp or expansion maps have any definition of roofs apart from the one inherent in the terrain for the lower level as far as I've seen (and I think I've been over all of the base game ones several times by now). I'm also not aware of any way to define base camp or expansion construction on any level other than the level of the base camp itself. The coordinates specified are X and Y coordinates. If the syntax allows for an optional Z coordinate I'm unaware of it.

It can also be noted that the current terrain format in the game unfortunately defines the ceiling/roof as part of the floor/wall/... tile. I would love to see that broken up so you could mix your choice of floors/walls/... with whatever roof material you wanted specified on the next level (with the corresponding capability to actually define the next level for base camp/expansion construction, of course).

**Comment by Procyonae** [comment]
> No base camp or expansion maps have any definition of roofs apart from the one inherent in the terrain for the lower level as far as I've seen (and I think I've been over all of the base game ones several times by now). I'm also not aware of any way to define base camp or expansion construction on any level other than the level of the base camp itself. The coordinates specified are X and Y coordinates. If the syntax allows for an optional Z coordinate I'm unaware of it.

The ability to do that should be added then, are there no multi storey base camp buildings rn? If not that seems like good reason to add it even if auto roofs did work reliably.
Auto roofs have been borked since their inception as far as I'm aware and aren't being actively supported anymore to my knowledge due to almost every map in the game having a manually defined one (The only maps I can think of rn is the FEMA camp which ideally needs making into a mutable at some point to get around the lack of z-level nests and the hardcoded lab tower which I plan on unhardcoding soonish)
I'd happily chip away at adding manual roofs for all the basecamp stuff if the infra got added.

**Comment by PatrikLundell** [comment]
There are base camps in pre existing multi story buildings, but I think all of them are on the ground level, apart from the bare bones base camp which can be placed anywhere, but at the cost of not being able to place any expansions when not at Z level 0 (as the expansions would end up underground or in the air/on roof tops, where no general expansions exist).

Note that we're still hobbled by the mashed floor/roof terrain definition, and the construction costs are based on the roofs specified in the terrain descriptions are included in the time and materials estimates. This doesn't matter for ordinary map generation, as neither time nor materials parameters are used.

What we'd really need is:
- Terrain tiles specifying what's on their Z level only. This means the inherent data structure as well as all the definitions that now include roofs.
- Means for players to "manually" (construction menu, zone construction by PC or companion) construct things at Z level 1 (scaffolding?, step ladders?, as well as the ability to construct from an adjacent tile once you've got the first floor/roof build on the next level). This also means logic that "understands" how to construct things on the next Z level for the zone construction case (pathfinding and collection/set up/take down equipment needed for access, although I guess dedicated scaffolding could be defined, so you construct the scaffolding first and then whatever you want to reach from it. Note that you may want the scaffolding to remain if you want to add another level on top of the current one).
- Updates of the JSON format for base camp construction allowing for the specification of the Z level. This change is actually independent of the ones above, in that it can be used immediately using the current crummy mashed together terrain definitions and then "just" (lots JSON file adjustments to replace the roof components, but that can mostly be done via palette replacement, plus the much more laborious task of actually creating new entries for each and every roof).
- Updates to the base camp expansion logic so it can handle other Z levels. The first primitive step would be to have expansions automatically end up Z level 0, with a second step being to allow you to select whether to expand on the current level or Z level 0 (when different), a further one being to ability to expand in any of the 26 available tiles on Z level current -1, current level, and current level +1, and a potential "final" one being able to expand to +/- the reality bubble Z level coverage.

However, I suspect we need auto roof to work for base camps until the pipe dream above has been implemented by someone competent (i.e. not me).

**Comment by Procyonae** [comment]
> and the construction costs are based on the roofs specified in the terrain descriptions are included in the time and materials estimates

That's kind of weird bc individually constructing walls gives roofs for free, uniformity in how it works would be good

> * Terrain tiles specifying what's on their Z level only. This means the inherent data structure as well as all the definitions that now include roofs.

Agreed, I hate the way this is done currently

> * Means for players to "manually" (construction menu, zone construction by PC or companion) construct things at Z level 1 (scaffolding?, step ladders?, as well as the ability to construct from an adjacent tile once you've got the first floor/roof build on the next level). This also means logic that "understands" how to construct things on the next Z level for the zone construction case (pathfinding and collection/set up/take down equipment needed for access, although I guess dedicated scaffolding could be defined, so you construct the scaffolding first and then whatever you want to reach from it. Note that you may want the scaffolding to remain if you want to add another level on top of the current one).

For manual construction this probably isn't even that difficult from what I've tampered with in the construction code, the main difficulty might be adding the ability to "peek" up and down while placing stuff so you can see what you're doing and that would hopefully just be reusing code

> However, I suspect we need auto roof to work for base camps until the pipe dream above has been implemented by someone competent (i.e. not me).

I'd like to think it's not a pipe dream, with the first two points not being necessary and the third hopefully being reasonably easy (as in the functionality I realise adding the roofs would be alot of work, regex replacement of the lower level might help to get the shape to cut down the time it takes but either way I'd happily do this) and the fourth step isn't really necessary for roofs specifically either, you'd just have them be built when they currently are.

I'll have a look at the code side of the third at some point and see if it's more complicated than it seems given that I haven't touched basecamp code yet. I think I could handle making roof constructions work from scaffolding etc but probably not the vision side of things and it probably wouldn't be sensible to merge without that unless I just did the infra but didn't touch the existing way it works in the hopes that would convince someone else to do the vision part. I'd be happy to audit and separate all the constructions to work this way if it does get done.

**Comment by PatrikLundell** [comment]
I think you misunderstood me in the case of how the current system works. Constructing a wall "manually" also constructs the roof, and the construction cost reflects that. Constructing it as a base camp construction task costs the same, but the roof is often/usually missing.

If you were to construct the roof as part of the wall construction you'd need to select both the wall/floor/... and whatever ceiling/roof you want somehow, so it's two selections (and it gets worse if you want to place real world restrictions on it, such as not placing a reinforced concrete roof on top of an earth wall).

For the "manual" construction case I'm not sure it's ideal. The logic handles you building a pit, half a log wall, and then a full log wall as three chained tasks, and technically chaining a roof selection on top if that is probably comparatively easy, but you'd get a combinatorial issue with the wall + roof selection (in particular if you add the logical extension of building a wall on top of the wall for another story).

Also note that we seem to be gearing up towards a new "full" release, so this kind of work is likely to get put on hold until after the release and subsequent catch up. That doesn't mean work can't be done, but it may not be merged, and when it's finally time to merge it there's a significant risk you'd have to deal with merge conflicts before it would be accepted for review.

**Comment by Procyonae** [comment]
> I think you misunderstood me in the case of how the current system works. Constructing a wall "manually" also constructs the roof, and the construction cost reflects that. Constructing it as a base camp construction task costs the same, but the roof is often/usually missing.

Oh neat I didn't realise the manual construction factored the cost in

> If you were to construct the roof as part of the wall construction you'd need to select both the wall/floor/... and whatever ceiling/roof you want somehow, so it's two selections (and it gets worse if you want to place real world restrictions on it, such as not placing a reinforced concrete roof on top of an earth wall).

If you're referring to basecamps I don't see why picking a roof material adds any complication personally, it'd just be a single terrain pick unlike the existing basecamp palettes.
For manual constructions you'd just build the roof afterwards.
I fail to see how this relates to real world restrictions given they're not handled at all currently.

> For the "manual" construction case ... but you'd get a combinatorial issue with the wall + roof selection (in particular if you add the logical extension of building a wall on top of the wall for another story).

I don't get what you mean by this. Given that all our roofs are flat currently I don't see why we can't allow building walls and surfacing them into internal floors if that's relevant. This can always be refined down the line.
 
> Also note that we seem to be gearing up towards a new "full" release, so this kind of work is likely to get put on hold until after the release and subsequent catch up. That doesn't mean work can't be done, but it may not be merged, and when it's finally time to merge it there's a significant risk you'd have to deal with merge conflicts before it would be accepted for review.

I thought we weren't doing a feature freeze for 0.H though so why would it get put on hold? Just peeps working on triage instead or?

**Comment by Procyonae** [comment]
I'm also confused why this has been added to 0.H blockers, issues with auto roofs long predate 0.G to my knowledge so regardless how this is solved it seems unlikely to be fixed before the next stable.

**Comment by PatrikLundell** [comment]
Last comment first: It's a bot that added it to blockers, so I assume it's a result of it being added to triage, with the result of that deciding whether to take it off or not. Also note that we're discussing something a bit more complicated than fixing the bug this report really covers (assuming it's easier to get the missing roofs generated than splitting the terrain tile ground and ceiling parts and moving the latter to the next Z level).

If "manual" construction would add the roof material in a separate step, then base camp construction should as well. Note that a single "job" can involve multiple sets of constructions, so assuming there was support for constructing it at the level above was implemented (the hard part) it's just a matter of adding a lot of additional roof maps to crown those walls. Perspiration, but not inspiration.

The current terrain tiles include roofs of "appropriate" materials to go with the walls. The simple solution would be to un-bundle these sets and introduce ceiling/roof tiles that can go with any wall, whereas a more complicated solution would have restrictions on what kind of walls you can have under your roof (similar to how different kinds of terrain restrict what you can construct on them, with the wall on the level below being the terrain). While all roofs currently are flat, they're still made of different materials (allocated when constructing the tiles), at least if you construct them (generated maps might be different: I haven't looked into that side). There's at least the white roof, a wooden roof, and the shingle roofs (plus sky lights, greenhouse roofs, etc.) visible at the level above. I don't think you can construct walls with gutters on top of them, to mention one thing map generation can do that construction can't (essentially making an invisible ceiling (I don't think there's any way to view the ceiling: you can only see the top side, i.e. the roof/floor on the next level) made from a different material than the roof that forms the other side of that ceiling).

I didn't know there wouldn't be a feature freeze this time. Sure, it was a fair bit too long last time, but you don't really want to make a new major version with half of a new feature in it. On the other hand, we don't see any such activities for stable versions, so maybe it can be handled behind the scenes.

**Comment by Procyonae** [comment]
> If "manual" construction would add the roof material in a separate step, then base camp construction should as well. Note that a single "job" can involve multiple sets of constructions, so assuming there was support for constructing it at the level above was implemented (the hard part) it's just a matter of adding a lot of additional roof maps to crown those walls. Perspiration, but not inspiration.

I appreciate that

> The current terrain tiles include roofs of "appropriate" materials to go with the walls. The simple solution would be to un-bundle these sets and introduce ceiling/roof tiles that can go with any wall, whereas a more complicated solution would have restrictions on what kind of walls you can have under your roof (similar to how different kinds of terrain restrict what you can construct on them, with the wall on the level below being the terrain)

Limiting roof types per wall is simple enough but how would you go about preventing non wall roofs from being different types? If you make it so there has to be the same type adjacent then you'd lose the ability to mix and match roofs and mean stuff like skylights would have to be dealt with (would they be allowed with any roof? would the roofs and walls be split into tiers with each wall tier supporting more roof tiers and higher roof tiers allowing skylights etc?)
I thought currently you can construct whichever roof you like after the walls but I don't have time to check rn, either way the simple solution wouldn't be that big a downgrade and wouldn't make the complicated way any more work in future.

> I don't think you can construct walls with gutters on top of them

Probably not the construction menu is missing quite alot, ideally gutter would be furniture anyway which I'll probably get around to at some point now they're only using a single terrain id across the whole repo (once #70584 is merged at least)

> essentially making an invisible ceiling (I don't think there's any way to view the ceiling: you can only see the top side, i.e. the roof/floor on the next level) made from a different material than the roof that forms the other side of that ceiling

Eh? I don't get what you're trying to say here at all, when can you normally "view" the ceiling?

> I didn't know there wouldn't be a feature freeze this time. Sure, it was a fair bit too long last time, but you don't really want to make a new major version with half of a new feature in it. On the other hand, we don't see any such activities for stable versions, so maybe it can be handled behind the scenes.

From what I've gleamed someone's going to pick a commit to base 0.H off of and then bugfixes etc will be backported more akin to how the point releases have worked but idk, if you're a collaborator on the devcord apparently they've talked about it more in the collab+ channel

**Comment by Maleclypse** [comment]
No I added it to triage as something I'd like to see people work on.  I can guarantee that our triage lists are much more expansive than what ends up being used as a full blocker.  This will be the kind of item that's looked on as a "Neat to have, booted to 0.I blocker list" if no meaningful progress has started on it by the time we are close to calling 0.H done.

**Comment by PatrikLundell** [comment]
You'd limit roof types over floors either based on new logic depending on supporting roof tiles or by checking the tile of the floor below in a similar manner as is used to determine support currently, or a combination of both. The simple solution is indeed to just check for support and don't care about what kind of "roof" you construct ("roof" because it might the the floor on the next level instead).
It can also be noted that a roof/next floor being split away from the current terrain tile might tie into a desire to make support have a range (so you'd need support pillars to cover a large floor). If so, different "roof" materials may have differing abilities to carry support for different distances, resulting in a somewhat natural restriction. That might also be tied to wall strength. However, such logic should probably be introduced separately.

Making gutters into furniture sounds logical, although I think that would disallow the placement of birds nests on them.

I was trying to say that you can never see the ceiling. The only sort of exception is that some map generated buildings have lights in the ceiling, but I assume these are furniture (which I'd like to be able to construct "manually", but that's not in the scope of the current discussion).

**Comment by Procyonae** [comment]
> Making gutters into furniture sounds logical, although I think that would disallow the placement of birds nests on them.

Weirdly they're not even an item, the birdnest nest(ed map) spawns twigs, feathers and eggs and that's it.

> I was trying to say that you can never see the ceiling. The only sort of exception is that some map generated buildings have lights in the ceiling, but I assume these are furniture (which I'd like to be able to construct "manually", but that's not in the scope of the current discussion).

They're actually disgustingly terrain (https://cdda-guide.nornagon.net/terrain/t_floor_olight for example), and would need converting to appliances before they're constructible really but there's no support for "ceiling mounted" appliances yet (I don't think there's any "ceiling mounted" furniture either but that would be very easy to add)

**Comment by PatrikLundell** [comment]
I can see four structural ways to ensure construction of ceilings result in construction of roofs/floors visible on the next level (letters do not represent ranking):
A. Change the construction code to generate an additional tinymap for the level above and apply the appropriate terrain changes on that map.
B. Introduce a smallmap with the same footprint as the tinymap (24*24), but with Z levels active and use that instead of a tinymap when (potentially) dealing with roofs. This may or may not require additional code to get the roofs into place (the current code is weird in that usage of a different Z level with a tinymap just silently operates on the only supported Z level, which typically does nothing). It's possible the roof adding code that's currently failing silently with tinymap would work with a smallmap, but it's also possible not all the required code is there.
C. Give up on using tinymap and use map whenever a roof construction is a possibility. That's probably rather wasteful, given that map is much larger (132 * 132 tiles, all Z levels) than the area you're actually manipulating.
D. Change tinymap to support Z levels. That would make it 21 times the current size (1 Z level vs. 1 +/- 10) and thus have some performance impact. Also a potential for side effects if some usages rely on effects NOT carrying over to other Z levels, but that's really rather dodgy usage, if present.
